File: 2021_winter_algorithm_study_intermediate/DAY05_SEGMENT_TREE/Project1/Project1/19646.cpp
#include<iostream>

using namespace std;

int d[200001 * 4], e[200001];
int N, i, x;

int get(int p, int left, int right) {
    if (left == right) {
        return d[p] = e[left];
    }
    return d[p] = get(p * 2, left, (left + right) / 2) + get(p * 2 + 1, (left + right) / 2 + 1, right);
}

int f(int p, int left, int right, int v) {
    if (left == right) {
        return left;
    }
    if (d[p * 2] < v) {
        return f(p * 2 + 1, (left + right) / 2 + 1, right, v - d[p * 2]);
    }
    return f(p * 2, left, (left + right) / 2, v);
}


int update(int p, int left, int right, int x) {
    if (x < left || right < x) {
        return d[p];
    }
    if (left == right) {
        return d[p] = 0;
    }
    return d[p] = update(p * 2, left, (left + right) / 2, x) + update(p * 2 + 1, (left + right) / 2 + 1, right, x);
}

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);

    cin >> N;
    for (i = 1; i <= N; i++) {
        cin >> e[i];
    }
    get(1, 1, N);
    for (i = 0; i < N; i++) {
        cin >> x;
        x = f(1, 1, N, x);
        cout << x << " ";
        update(1, 1, N, x);
    }
}

File: 2021_winter_algorithm_study_intermediate/DAY05_SEGMENT_TREE/19646.py
import sys
sys.stdin = open("input.txt", "r")

N = int(input())
d = [0 for _ in range(4*N)]
e = [0] + list(map(int, input().split()))


def get(p, left, right):
    if left == right:
        d[p] = e[left]
        return d[p]
    d[p] = get(p * 2, left, (left + right) // 2) + get(p * 2 + 1, (left + right) // 2 + 1, right)
    return d[p]


def sol(p, left, right, v):
    if left == right:
        return left
    if d[p*2] < v:
        return sol(p * 2 + 1, (left + right) // 2 + 1, right, v - d[p * 2])
    return sol(p * 2, left, (left + right) // 2, v)


def update(p, left, right, x):
    if x < left or right < x:
        return d[p]
    if left == right:
        d[p] = 0
        return d[p]
    d[p] = update(p * 2, left, (left + right) // 2, x) + update(p * 2 + 1, (left + right) // 2 + 1, right, x)
    return d[p]


get(1, 1, N)
p = list(map(int, input().split()))

for i in range(N):
    p[i] = sol(1, 1, N, p[i])
    print(p[i], end=" ")
    update(1, 1, N, p[i])


